### PR TITLE
For tests purposes, adds a memory limiter under the form of a decorator.

### PR DESCRIPTION
**Pull request by AlexandreAbraham**
It is not used for the moment but was requested in #736.

### CODE REVIEW

File: nilearn/_utils/testing.py [hunk 1]
```diff
@@ -647,3 +649,20 @@ def assert_less_equal(a, b):
     def assert_less(a, b):
         if a >= b:
             raise AssertionError("%f is not less than %f" % (a, b))
+
+
+def MemoryLimit(limit):
```
Comment (GaelVaroquaux) [inline nilearn/_utils/testing.py]: CamelCase for a function. Blah!

File: nilearn/_utils/testing.py [hunk 2]
```diff
@@ -10,6 +10,8 @@
 import sys
 import tempfile
 import warnings
+import resource
```
Comment (GaelVaroquaux) [inline nilearn/_utils/testing.py]: Are you sure that this is in the standard library?

File: nilearn/_utils/testing.py (hunk 1 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/testing.py]: I use it as a decorator afterward. But I can change it to `snake_case` since nose does that with `with_setup`.

**Comment by AlexandreAbraham** [comment]
Ping @aabadie, apparently I stole your job. But I've done the easy part, the hardest part is to set it cleverly when needed :P.

File: nilearn/_utils/testing.py (hunk 1 above)
Comment (GaelVaroquaux) [inline nilearn/_utils/testing.py]: > I use it as a decorator afterward.

Decorators aRe nOt CamelCase. It's uGlY. :)

File: nilearn/_utils/testing.py (hunk 2 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/testing.py]: It is for Unix systems. I probably should skip tests if not available / setrlimit or getrlimit raise a resource.error.

Comment (GaelVaroquaux) [inline nilearn/_utils/testing.py]: > It is for Unix systems. I probably should skip tests if not available /
> setrlimit or getrlimit raise a resource.error.

More than skip tests, you cannot import it at all windows, I suspect.
Which means that importing scikit-learn on windows would crash.

We need CI for Windows. I created an issue: #804.

**Comment by aabadie** [comment]
> Ping @aabadie, apparently I stole your job. But I've done the easy part, the hardest part is to set it cleverly when needed :P.

No problem @AlexandreAbraham, your change is a nice contribution.

**Comment by aabadie** [comment]
@AlexandreAbraham,  in case you missed it, the test is failing on circle ci. Probably an issue when setting memory limit on the VM.

**Comment by AlexandreAbraham** [comment]
@aabadie, setting the limit of memory is protected. The error seems to be that a segfault is triggered when I allocate too much memory instead of a memory error. I'll try to put a smaller object in the test.

**Comment by lesteve** [comment]
> @aabadie, setting the limit of memory is protected. The error seems to be that a segfault is triggered when I allocate too much memory instead of a memory error. I'll try to put a smaller object in the test.

You put a smaller object but didn't change the memory_limit. I would guess that's why the Travis tests are now failing.

**Comment by AlexandreAbraham** [comment]
I don't get your comment. The error should be triggered when the allocated object size is higher than the limit, which is the case. Or maybe I have my memory usage estimation wrong?

**Comment by lesteve** [comment]
> I don't get your comment. The error should be triggered when the allocated object size is higher than the limit, which is the case. Or maybe I have my memory usage estimation wrong?

I haven't double-checked the memory usage computation but it looks like the Travis failure is due to the fact that the memory usage doesn't go over the memory_usage threshold.

**Comment by lesteve** [comment]
On top of that it looks like it only fails for a few distributions but not for the other, weird.

Here is the error you get:

```
======================================================================

FAIL: nilearn.tests.test_testing.test_memory_limit_raise

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest

    self.test(*self.arg)

  File "/home/travis/build/nilearn/nilearn/testvenv/local/lib/python2.7/site-packages/nilearn-0.1.5.dev0-py2.7.egg/nilearn/_utils/testing.py", line 690, in wrapper_func

    retval = func(*args, **kwargs)

  File "/home/travis/build/nilearn/nilearn/testvenv/local/lib/python2.7/site-packages/nilearn-0.1.5.dev0-py2.7.egg/nilearn/tests/test_testing.py", line 58, in test_memory_limit_raise

    MemoryError, '', np.zeros, (500, 100, 60), dtype=np.float32)

AssertionError: MemoryError not raised
```

**Comment by AlexandreAbraham** [comment]
I can read the error, yes :P. What I don't understand is why it pops up only on some configurations.

**Comment by AlexandreAbraham** [comment]
The problem seems that the code segfaults instead of raising a memory error... Maybe I asked for too much memory?

**Comment by aabadie** [comment]
> The problem seems that the code segfaults instead of raising a memory error... Maybe I asked for too much memory?

This is a possibility but I would be surprised as the maximum memory is set to 10000000 (10MB) and the size of the array is 12MB which is not that much.

Maybe using a strategy based on MemoryProfiler would help.

**Comment by GaelVaroquaux** [comment]
> The problem seems that the code segfaults instead of raising a memory error...
> Maybe I asked for too much memory?

Indeed, under old versions of Python, there were segfaults when you asked
for too much memory.

**Comment by lesteve** [comment]
FWIW @aabadie has a solution for this use case (testing nilearn memory consumption as per #736) based on memory_profiler, which is part of https://github.com/joblib/joblib/pull/260.

I am not entirely sure it is worth investigating more time making this PR work.

**Comment by aabadie** [comment]
> FWIW @aabadie has a solution for this use case (testing nilearn memory consumption as per #736) based on memory_profiler

 @AlexandreAbraham, @lesteve, if you agree, I'll propose a PR based on that solution.

**Comment by lesteve** [comment]
> @AlexandreAbraham, @lesteve, if you agree, I'll propose a PR based on that solution.

This is fine with me. It would be great to find a few tests to add in order to test functions in nilearn where  memory consumption actually matters. @AlexandreAbraham, @GaelVaroquaux any suggestions ?

**Comment by AlexandreAbraham** [comment]
> Indeed, under old versions of Python, there were segfaults when you asked
for too much memory.

But we don't use older versions of Python. This is why this seems strange to me.

> FWIW @aabadie has a solution for this use case (testing nilearn memory consumption as per #736) based on memory_profiler, which is part of joblib/joblib#260.

Yes, I know how to do it with `memory_profiler`. The goal of this PR was to rely on the system and avoid a dependency.

**Comment by AlexandreAbraham** [comment]
> @AlexandreAbraham, @lesteve, if you agree, I'll propose a PR based on that solution.

I won't say no :P

**Comment by AlexandreAbraham** [comment]
Handing it over to @aabadie !

**Pull request closed without merging**